### PR TITLE
doc: Add missed copyright headers

### PR DESCRIPTION
**Pull request by hebasto**
This PR improves `contrib/devtools/copyright_header.py` script and adds copyright headers to the files in `src` and `test` directories with two exceptions:
- [`src/reverse_iterator.h`](https://github.com/bitcoin/bitcoin/blob/master/src/reverse_iterator.h) (added to exceptions)
- [`src/test/fuzz/FuzzedDataProvider.h`](https://github.com/bitcoin/bitcoin/blob/master/src/test/fuzz/FuzzedDataProvider.h) (added to exceptions)

On master 5622d8f3156a293e61d0964c33d4b21d8c9fd5e0:
```
$ ./contrib/devtools/copyright_header.py report . | grep zero
  25 with zero copyrights
```

With this PR:
```
$ ./contrib/devtools/copyright_header.py report . | grep zero
   2 with zero copyrights
```

~I am uncertain about our copyright policy with `build_msvc` and `contrib` directories content, so they are out of scope of this PR.~

### CODE REVIEW

**Comment by practicalswift** [comment]
Concept ACK

While we are at it: perhaps it is time to do the switch from `http` to `https` for the `opensource.org` links too?

Alternatively we would just switch from …

```
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.
```

… to …

```
// Distributed under the MIT software license, see the accompanying
// file COPYING.
```

I think I prefer the latter: we shouldn't rely on external third party sites unless necessary :)

**Comment by fanquake** [comment]
Concept ACK on adding the missing headers. 

In the future I'd prefer if these scripts were moved out of `bitcoin/bitcoin`; maybe into the maintainer tools repository. To me it seems a bit ridiculous that we've got 600 lines of bash in this repository to do copyright header "management".

File: contrib/devtools/copyright_header.py [hunk 1]
```diff
@@ -491,17 +491,28 @@ def file_has_hashbang(file_lines):
         return False
     return file_lines[0][:2] == '#!'
 
+def insert_shell_header(filename, file_lines, start_year, end_year):
```
Comment (MarcoFalke) [inline contrib/devtools/copyright_header.py]: Why is this needed? Couldn't the function where you copied this from be used?

**Review by maflcko — APPROVED** [review]

**Comment by MarcoFalke** [comment]
>  To me it seems a bit ridiculous that we've got 600 lines of bash 

I think it is python, but still ridiculous. Though having it in the repo, at least allows to do it in a scripted-diff. For the future, I am hoping we can get rid of the years after (c).

File: contrib/devtools/copyright_header.py (hunk 1 above)
Comment (hebasto) [reply contrib/devtools/copyright_header.py]: Fixed.

**Comment by hebasto** [comment]
@practicalswift 
I'm not sure that the license header content changing should be addressed by this PR.

@MarcoFalke 
Your comment has been addressed.

@sipsorcery 
What do you think about copyright headers in these files:
- `build_msvc/bitcoin_config.h`
- `build_msvc/msvc-autogen.py`
- `build_msvc/testconsensus/testconsensus.cpp`
?

@dongcarl 
What do you think about copyright headers in these files:
- `contrib/guix/guix-build.sh`
- `contrib/guix/libexec/build.sh`
?

**Comment by sipsorcery** [comment]
@hebasto sounds like a good idea to me. There was no particular reason copyright headers were left out of the files.

**Comment by hebasto** [comment]
@sipsorcery 
> @hebasto sounds like a good idea to me. There was no particular reason copyright headers were left out of the files.

Thanks, missed headers are added in `build_msvc/` subdirectory in the latest push.

**Comment by MarcoFalke** [comment]
ACK fac86ac7b3ceac2f884412c7a9f4bd5bab5e3916
